pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,264 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+	//"encoding/json"  temporary for debug
+	"github.com/golang/glog"
+	. "github.com/kubevirt/containerized-data-importer/pkg/common"
+	"github.com/pkg/errors"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	// pvc annotations
+	AnnCloneRequest    = "k8s.io/CloneRequest"
+	AnnCloneOf         = "k8s.io/CloneOf"
+	AnnCloneInProgress = "cdi.kubevirt.io/storage.clone.inprogress"
+	
+)
+
+type CloneController struct {
+	clientset                kubernetes.Interface
+	pvcQueue, podQueue       workqueue.RateLimitingInterface
+	pvcInformer, podInformer cache.SharedIndexInformer
+	cloneImage               string
+	pullPolicy               string // Options: IfNotPresent, Always, or Never
+	verbose                  string // verbose levels: 1, 2, ...
+}
+
+func NewCloneController(client kubernetes.Interface, pvcInformer, podInformer cache.SharedIndexInformer, cloneImage string, pullPolicy string, verbose string) *CloneController {
+	c := &CloneController{
+		clientset:     client,
+		pvcQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		podQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		pvcInformer:   pvcInformer,
+		podInformer:   podInformer,
+		cloneImage:    cloneImage,
+		pullPolicy:    pullPolicy,
+		verbose:       verbose,
+	}
+
+	// Bind the pvc SharedIndexInformer to the pvc queue
+	c.pvcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+		// this is triggered by an update or it will also be
+		// be triggered periodically even if no changes were made.
+		UpdateFunc: func(old, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Bind the pod SharedIndexInformer to the pod queue
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
we need a delete function here. If the pod get's deleted, we need to re-enqueue processing the pvc.

Reply:
this would also apply to the importer controller. Maybe both can be corrected in a future pr?